src/Google/GlobalSiteTag.php

```diff
@@ -155,13 +191,223 @@ public function maybe_display_event_snippet( string $ads_conversion_id, string $
+			esc_js( $order->get_currency() ),
+			esc_js( $this->wc->get_base_country() ),
+			esc_js( $order->get_total() ),
+			esc_js( $is_new_customer ),
+			esc_js( $order->get_cart_tax() ),
+			esc_js( $order->get_total_shipping() ),
+			esc_js( $order->get_shipping_postcode() ),
+			esc_js( $this->wc->get_base_country() ),
+			esc_js( $language ),
+		);
+		wp_print_inline_script_tag( $purchase_page_gtag );
+	}
+
+	/**
+	 * Display the JavaScript code to track the product view page.
+	 */
+	private function display_view_item_event_snippet(): void {
+		// Only display on the product view page.
+		if ( ! is_product() ) {
+			return;
+		}
+		$product        = wc_get_product( get_the_ID() );
+		$view_item_gtag = sprintf(
+			'gtag("event", "view_item", {
+                    "send_to": "GLA",
+                    "developer_id.%s": "true",
+                    "ecomm_pagetype": "product",
+                    "value": "%s",
+                    items:[{
+                    "id": "gla_%s",
+                    "price": %s,
+                    "google_business_vertical": "retail",
+                    "name":"%s",
+                    "category":"%s",
+                    }]});',
+			esc_js( self::DEVELOPER_ID ),
+			esc_js( (string) $product->get_price() ),
+			esc_js( $product->get_id() ),
+			esc_js( (string) $product->get_price() ),
+			esc_js( $product->get_name() ),
+			esc_js( join( '& ', $this->product_helper->get_categories( $product ) ) ),
+		);
+		wp_print_inline_script_tag( $view_item_gtag );
+	}
+
+	/**
+	 * Display the JavaScript code to track all pages.
+	 */
+	private function display_page_view_event_snippet(): void {
+		$page_view_gtag =
+		sprintf(
+			'gtag(
+                    "event", "page_view", {
+                    "send_to": "GLA",
+                    "developer_id.%s": "true",});',
+			esc_js( self::DEVELOPER_ID )
+		);
+		wp_print_inline_script_tag( $page_view_gtag );
+	}
+
+	/**
+	 * Display the JavaScript code to track the cart page.
+	 */
+	private function display_cart_page_snippet(): void {
+		// Only display on the cart page.
+		if ( ! is_cart() ) {
+			return;
+		}
+
+		$item_info = '';
+
+		foreach ( WC()->cart->get_cart() as $cart_item ) {
+			// gets the product id
+			$id = $cart_item['product_id'];
+
+			// gets the product object
+			$product = $cart_item['data'];
+			$name    = $product->get_name();
+			$price   = $product->get_price();
+			// gets the cart item quantity
+			$quantity = $cart_item['quantity'];
+
+			$item_info = $item_info . sprintf(
+				'{
+				"id": "gla_%s",
+				"price": %s,
+				"google_business_vertical": "retail",
+				"name":"%s",
+				"quantity":"%s",
+				}',
+				esc_js( $id ),
+				esc_js( $price ),
+				esc_js( $name ),
+				esc_js( $quantity )
+			);
+		}
+		$value          = WC()->cart->total;
+		$page_view_gtag = sprintf(
+			'gtag("event", "page_view",
+				{"send_to": "GLA",
+				"ecomm_pagetype": "cart",
+				"value": "%s",
+				 items: [' . $item_info . ']});',
```

Comment:
Since we are using sprintf here we can just replace this with a `%s` and add `$item_info` to the list of variables. I guess it's a matter of personal preference but I'd suggest to avoid mixing different methods to replace variables.

Personally I find this the clearest so you don't have to look which %s matches which variable, but sprintf is perfectly fine as well:

```php
		$page_view_gtag = "
			gtag(
				'event',
				'page_view',
				{
					send_to: 'GLA',
					ecomm_pagetype: 'cart',
					value: {$value},
					items: [
						{$item_info}
					]
				}
			);
		";
```

Reply:
Updated.